kbase-extension/static/kbase/js/common/jobManager.js

```diff
@@ -123,217 +296,326 @@ define(['common/html', 'common/jobs', 'common/ui', 'util/string'], (html, Jobs,
+             */
+            executeActionByJobStatus(args) {
+                const { action, statusList, validStatuses } = args;
+
+                const jobIdList = this.getJobIDsByStatus(statusList, validStatuses);
+                if (!jobIdList || !jobIdList.length) {
+                    return Promise.resolve(false);
+                }
+
+                return JobMessages.showDialog({ action, statusList, jobIdList }).then(
+                    (confirmed) => {
+                        if (confirmed) {
+                            this.doJobAction(action, jobIdList);
+                        }
+                        return confirmed;
                     }
+                );
+            }
+
+            /**
+             * Cancel a single job from the batch
+             *
+             * @param {string} jobId
+             */
+            cancelJob(jobId) {
+                return this.executeActionOnJobId({ jobId, action: 'cancel' });
+            }
+
+            /**
+             * Retry a single job from the batch
+             *
+             * @param {string} jobId
+             */
+            retryJob(jobId) {
+                return this.executeActionOnJobId({ jobId, action: 'retry' });
+            }
+
+            /**
+             * Cancel all jobs with the specified statuses
+             *
+             * @param {array} statusList - array of statuses to cancel
+             */
+            cancelJobsByStatus(statusList) {
+                return this.executeActionByJobStatus({
+                    action: 'cancel',
+                    statusList: statusList,
+                    validStatuses: Jobs.validStatusesForAction.cancel,
+                });
+            }
+
+            /**
+             * Retry all jobs that ended with the specified status(es)
+             * Although only jobs in status 'error' or 'terminated' can be retried,
+             * the frontend job status may be out of date, so it is left to ee2 to
+             * verify whether a job can be retried or not.
+             *
+             * @param {array} statusList - array of statuses to retry
+             */
+            retryJobsByStatus(statusList) {
+                return this.executeActionByJobStatus({
+                    action: 'retry',
+                    statusList: statusList,
+                    validStatuses: Jobs.validStatusesForAction.retry,
+                });
+            }
+
+            /**
+             * Cancel or retry a list of jobs
+             *
+             * @param {string} action - either 'cancel' or 'retry'
+             * @param {array} jobIdList
+             */
+            doJobAction(action, jobIdList) {
+                this.bus.emit(jobCommand[action].command, {
+                    jobIdList,
+                });
+                // add the appropriate listener
+                this.addListener(jobCommand[action].listener, jobIdList);
+            }
+        };
+
+    const BatchInitMixin = (Base) =>
```

Comment:
Mixin providing functions for bulk cells

Reply:
Interesting - will this also (eventually) export a JobManager variant for the single app cell, as opposed to the batch one?